### PR TITLE
Update rl_replacements.py returned hidden states from logprobs

### PR DESCRIPTION
**Pull request by pluesclues**

### CODE REVIEW

File: unsloth/models/rl_replacements.py [hunk 1]
```diff
@@ -214,18 +214,34 @@ def _get_per_token_logps(self, model, input_ids, attention_mask, logits_to_keep,
...
-            logits = logits.to(torch.float32)
+            hidden_states = model(input_ids=input_ids, attention_mask=attention_mask, logits_to_keep=logits_to_keep + 1).logits
+            #breakpoint()
+            #logits = logits[:, :-1, :]  # (B, L-1, V), exclude the last logit: it corresponds to the next token pred
+            return hidden_states
             input_ids = input_ids[:, -logits_to_keep:]
             # For transformers<=4.48, logits_to_keep argument isn't supported, so here we drop logits ourselves.
             # See https://github.com/huggingface/trl/issues/2770
             logits = logits[:, -logits_to_keep:]
+            logits = logits.to(torch.float32)
```
Comment (Datta0) [inline unsloth/models/rl_replacements.py]: NIT: Please remove this

File: unsloth/models/rl_replacements.py [hunk 2]
```diff
@@ -214,18 +214,34 @@ def _get_per_token_logps(self, model, input_ids, attention_mask, logits_to_keep,
...
+        os.environ["UNSLOTH_RETURN_HIDDEN_STATES"] = "1"
         with torch.amp.autocast(device_type = 'cuda', dtype = self._autocast_dtype):
             # We add 1 to `logits_to_keep` because the last logits of the sequence is later excluded
-            logits = model(input_ids=input_ids, attention_mask=attention_mask, logits_to_keep=logits_to_keep + 1).logits
-            logits = logits[:, :-1, :]  # (B, L-1, V), exclude the last logit: it corresponds to the next token pred
-            logits = logits.to(torch.float32)
+            hidden_states = model(input_ids=input_ids, attention_mask=attention_mask, logits_to_keep=logits_to_keep + 1).logits
+            #breakpoint()
+            #logits = logits[:, :-1, :]  # (B, L-1, V), exclude the last logit: it corresponds to the next token pred
+            return hidden_states
```
Comment (Datta0) [inline unsloth/models/rl_replacements.py:223]: If we're returning here, we don't need the code below. Remove perhaps?

File: unsloth/models/rl_replacements.py [hunk 3]
```diff
@@ -214,18 +214,28 @@ def _get_per_token_logps(self, model, input_ids, attention_mask, logits_to_keep,
...
             self._autocast_dtype = torch.float16 if os.environ.get('ACCELERATE_MIXED_PRECISION', 'fp16') == 'fp16' else torch.bfloat16
             if os.environ.get('UNSLOTH_FORCE_FLOAT32', '0') == '1': self._autocast_dtype = torch.float16
-        os.environ["UNSLOTH_RETURN_HIDDEN_STATES"] = "0"
+        os.environ["UNSLOTH_RETURN_HIDDEN_STATES"] = "1"
         with torch.amp.autocast(device_type = 'cuda', dtype = self._autocast_dtype):
             # We add 1 to `logits_to_keep` because the last logits of the sequence is later excluded
-            logits = model(input_ids=input_ids, attention_mask=attention_mask, logits_to_keep=logits_to_keep + 1).logits
-            logits = logits[:, :-1, :]  # (B, L-1, V), exclude the last logit: it corresponds to the next token pred
-            logits = logits.to(torch.float32)
+            hidden_states = model(input_ids=input_ids, attention_mask=attention_mask, logits_to_keep=logits_to_keep + 1).logits
```
Comment (Datta0) [inline unsloth/models/rl_replacements.py:220]: NIT: These are logits. can we rename appropriately

File: unsloth/models/rl_replacements.py (hunk 2 above)
Comment (pluesclues) [reply unsloth/models/rl_replacements.py:223]: I want to leave it there for debugging purposes, dan had the old code below because of that

File: unsloth/models/rl_replacements.py (hunk 3 above)
Comment (pluesclues) [reply unsloth/models/rl_replacements.py:220]: I actually return those as hidden states so we save memory and can calculate logits on the fly over here:

https://github.com/pluesclues/unsloth-zoo/blob/8b5bfe233f819aac89876025d121d6af28713af6/unsloth_zoo/rl_replacements.py#L127-L129
